man/rdiscount.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "RDISCOUNT" "1" "July 2011" "" "RUBY"
.
.SH "NAME"
\fBrdiscount\fR \- humane markup to HTML conversion tool
.
.SH "SYNOPSIS"
[\fBenv\fR RDISCOUNT_EXTENSIONS=\'\fIextension\fR,\.\.\.\'] \fBrdiscount\fR [\fIfile\fR\.\.\.]
.
.SH "DESCRIPTION"
The \fBrdiscount\fR utility reads one or more markdown(7)\-formatted text \fIfile\fRs and writes HTML to standard output\. With no \fIfile\fR, or when \fIfile\fR is \'\-\', \fBrdiscount\fR reads source text from standard input\.
.
.P
The \fBRDISCOUNT_EXTENSIONS\fR environment variable can specify a comma\-separated list of Markdown processing \fIextension\fRs to enable (see \fBEXTENSIONS\fR below)\.
.
.SH "EXTENSIONS"
.
.TP
\fBfilter_styles\fR
Do not output \fB<style>\fR tags\.
.
.TP
\fBfold_lines\fR
RedCloth compatible line folding (not used)\.
.
.TP
\fBMKD_NOLINKS\fR
Don\'t do link processing, block \fB<a>\fR tags
.
.TP
\fBMKD_NOIMAGE\fR
Don\'t do image processing, block \fB<img>\fR
.
.TP
\fBMKD_NOPANTS\fR
Don\'t run smartypants()
.
.TP
\fBMKD_NOHTML\fR
Don\'t allow raw html through AT ALL
.
.TP
\fBMKD_STRICT\fR
Disable SUPERSCRIPT, RELAXED_EMPHASIS
.
.TP
\fBMKD_TAGTEXT\fR
Process text inside an html tag; no \fB<em>\fR, no \fB<bold>\fR, no html or [] expansion
.
.TP
\fBMKD_NO_EXT\fR
Don\'t allow pseudo\-protocols
.
.TP
\fBMKD_CDATA\fR
Generate code for xml ![CDATA[\.\.\.]]
.
.TP
\fBMKD_NOSUPERSCRIPT\fR
No A
.
.TP
\fBMKD_NORELAXED\fR
Emphasis happens everywhere
.
.TP
\fBMKD_NOTABLES\fR
Don\'t process PHP Markdown Extra tables\.
.
.TP
\fBMKD_NOSTRIKETHROUGH\fR
Forbid
.
.TP
\fBMKD_TOC\fR
Do table\-of\-contents processing
.
.TP
\fBMKD_1_COMPAT\fR
Compatability with MarkdownTest_1\.0
.
.TP
\fBMKD_AUTOLINK\fR
Make http://foo\.com a link even without \fB<>\fRs
.
.TP
\fBMKD_SAFELINK\fR
Paranoid check for link protocol
.
.TP
\fBMKD_NOHEADER\fR
Don\'t process document headers
.
.TP
\fBMKD_TABSTOP\fR
Expand tabs to 4 spaces
.
.TP
\fBMKD_NODIVQUOTE\fR
Forbid >%class% blocks
.
.TP
\fBMKD_NOALPHALIST\fR
Forbid alphabetic lists
.
.TP
\fBMKD_NODLIST\fR
Forbid definition lists
.
.SH "RETURN VALUES"
The \fBrdiscount\fR utility exits 0 on success, and > 0 if an error occurs\.
.
.SH "SEE ALSO"
markdown(7), env(1)
.
.SH "AUTHOR"
Ryan Tomayko \fIhttp://tomayko\.com/about\fR


ext/rdiscount.c
#include <stdio.h>
#include <locale.h>
#include "ruby.h"
#include "mkdio.h"

typedef struct {
    char *accessor_name;
    int flag;
} AccessorFlagPair;

/* 
 * Maps accessor names on the RDiscount object to Discount flags.
 * 
 * The following flags are handled specially:
 * - MKD_TABSTOP: Always set.
 * - MKD_NOHEADER: Always set.
 * - MKD_DLEXTRA: Always set. (For compatibility with RDiscount 2.1.8 and earlier.)
 * - MKD_FENCEDCODE: Always set. (For compatibility with RDiscount 2.1.8 and earlier.)
 * - MKD_GITHUBTAGS: Always set. (For compatibility with RDiscount 2.1.8 and earlier.)
 * - MKD_NOPANTS: Set unless the "smart" accessor returns true.
 * 
 * See rb_rdiscount__get_flags() for the detailed implementation.
 */
static AccessorFlagPair ACCESSOR_2_FLAG[] = {
    { "filter_html", MKD_NOHTML },
    { "footnotes", MKD_EXTRA_FOOTNOTE },
    { "generate_toc", MKD_TOC },
    { "no_image", MKD_NOIMAGE },
    { "no_links", MKD_NOLINKS },
    { "no_tables", MKD_NOTABLES },
    { "strict", MKD_STRICT },
    { "autolink", MKD_AUTOLINK },
    { "safelink", MKD_SAFELINK },
    { "no_pseudo_protocols", MKD_NO_EXT },
    { "no_superscript", MKD_NOSUPERSCRIPT },
    { "no_strikethrough", MKD_NOSTRIKETHROUGH },
    { NULL, 0 }     /* sentinel */
};

static VALUE rb_cRDiscount;

int rb_rdiscount__get_flags(VALUE ruby_obj)
{
    AccessorFlagPair *entry;
    
    /* compile flags */
    int flags = MKD_TABSTOP | MKD_NOHEADER | MKD_DLEXTRA | MKD_FENCEDCODE | MKD_GITHUBTAGS;
    
    /* The "smart" accessor turns OFF the MKD_NOPANTS flag. */
    if ( rb_funcall(ruby_obj, rb_intern("smart"), 0) != Qtrue ) {
        flags = flags | MKD_NOPANTS;
    }
    
    /* Handle standard flags declared in ACCESSOR_2_FLAG */
    for ( entry = ACCESSOR_2_FLAG; entry->accessor_name; entry++ ) {
        if ( rb_funcall(ruby_obj, rb_intern(entry->accessor_name), 0) == Qtrue ) {
            flags = flags | entry->flag;
        }
    }
    
    return flags;
}

static VALUE
rb_rdiscount_to_html(int argc, VALUE *argv, VALUE self)
{
    /* grab char pointer to markdown input text */
    char *res;
    int szres;
    VALUE encoding;
    VALUE text = rb_funcall(self, rb_intern("text"), 0);
    VALUE buf = rb_str_buf_new(1024);
    Check_Type(text, T_STRING);

    int flags = rb_rdiscount__get_flags(self);
    
    /* 
     * Force Discount to use ASCII character encoding for isalnum(), isalpha(),
     * and similar functions.
     * 
     * Ruby tends to use UTF-8 encoding, which is ill-defined for these
     * functions since they expect 8-bit codepoints (and UTF-8 has codepoints
     * of at least 21 bits).
     */
    char *old_locale = strdup(setlocale(LC_CTYPE, NULL));
    setlocale(LC_CTYPE, "C");   /* ASCII (and passthru characters > 127) */

    MMIOT *doc = mkd_string(RSTRING_PTR(text), RSTRING_LEN(text), flags);

    if ( mkd_compile(doc, flags) ) {
        szres = mkd_document(doc, &res);

        if ( szres != EOF ) {
            rb_str_cat(buf, res, szres);
            rb_str_cat(buf, "\n", 1);
        }
    }
    mkd_cleanup(doc);

    setlocale(LC_CTYPE, old_locale);
    free(old_locale);

    /* force the input encoding */
    if ( rb_respond_to(text, rb_intern("encoding")) ) {
        encoding = rb_funcall(text, rb_intern("encoding"), 0);
        rb_funcall(buf, rb_intern("force_encoding"), 1, encoding);
    }

    return buf;
}

static VALUE
rb_rdiscount_toc_content(int argc, VALUE *argv, VALUE self)
{
    char *res;
    int szres;

    int flags = rb_rdiscount__get_flags(self);

    /* grab char pointer to markdown input text */
    VALUE text = rb_funcall(self, rb_intern("text"), 0);
    Check_Type(text, T_STRING);

    /* allocate a ruby string buffer and wrap it in a stream */
    VALUE buf = rb_str_buf_new(4096);

    MMIOT *doc = mkd_string(RSTRING_PTR(text), RSTRING_LEN(text), flags);

    if ( mkd_compile(doc, flags) ) {
        szres = mkd_toc(doc, &res);

        if ( szres != EOF ) {
            rb_str_cat(buf, res, szres);
            rb_str_cat(buf, "\n", 1);
        }
    }
    mkd_cleanup(doc);

    return buf;
}

int rb_rdiscount__get_flags(VALUE ruby_obj)
{
    VALUE flags = rb_funcall(ruby_obj, rb_intern("flags"), 0);
    return NUM2INT(rb_funcall(flags, rb_intern("to_i"), 0));
}

void Init_rdiscount()
{
    rb_cRDiscount = rb_define_class("RDiscount", rb_cObject);
    rb_define_const(rb_cRDiscount, "MKD_NOLINKS", INT2NUM(MKD_NOLINKS));
    rb_define_const(rb_cRDiscount, "MKD_NOIMAGE", INT2NUM(MKD_NOIMAGE));
    rb_define_const(rb_cRDiscount, "MKD_NOPANTS", INT2NUM(MKD_NOPANTS));
    rb_define_const(rb_cRDiscount, "MKD_NOHTML", INT2NUM(MKD_NOHTML));
    rb_define_const(rb_cRDiscount, "MKD_STRICT", INT2NUM(MKD_STRICT));
    rb_define_const(rb_cRDiscount, "MKD_TAGTEXT", INT2NUM(MKD_TAGTEXT));
    rb_define_const(rb_cRDiscount, "MKD_NO_EXT", INT2NUM(MKD_NO_EXT));
    rb_define_const(rb_cRDiscount, "MKD_CDATA", INT2NUM(MKD_CDATA));
    rb_define_const(rb_cRDiscount, "MKD_NOSUPERSCRIPT", INT2NUM(MKD_NOSUPERSCRIPT));
    rb_define_const(rb_cRDiscount, "MKD_NORELAXED", INT2NUM(MKD_NORELAXED));
    rb_define_const(rb_cRDiscount, "MKD_NOTABLES", INT2NUM(MKD_NOTABLES));
    rb_define_const(rb_cRDiscount, "MKD_NOSTRIKETHROUGH", INT2NUM(MKD_NOSTRIKETHROUGH));
    rb_define_const(rb_cRDiscount, "MKD_TOC", INT2NUM(MKD_TOC));
    rb_define_const(rb_cRDiscount, "MKD_1_COMPAT", INT2NUM(MKD_1_COMPAT));
    rb_define_const(rb_cRDiscount, "MKD_AUTOLINK", INT2NUM(MKD_AUTOLINK));
    rb_define_const(rb_cRDiscount, "MKD_SAFELINK", INT2NUM(MKD_SAFELINK));
    rb_define_const(rb_cRDiscount, "MKD_NOHEADER", INT2NUM(MKD_NOHEADER));
    rb_define_const(rb_cRDiscount, "MKD_TABSTOP", INT2NUM(MKD_TABSTOP));
    rb_define_const(rb_cRDiscount, "MKD_NODIVQUOTE", INT2NUM(MKD_NODIVQUOTE));
    rb_define_const(rb_cRDiscount, "MKD_NOALPHALIST", INT2NUM(MKD_NOALPHALIST));
    rb_define_const(rb_cRDiscount, "MKD_NODLIST", INT2NUM(MKD_NODLIST));
    rb_define_method(rb_cRDiscount, "to_html", rb_rdiscount_to_html, -1);
    rb_define_method(rb_cRDiscount, "toc_content", rb_rdiscount_toc_content, -1);
}

/* vim: set ts=4 sw=4: */
